install/0000_90_machine-config-operator_01_prometheus-rules.yaml

```diff
@@ -57,3 +57,23 @@ spec:
             severity: warning
           annotations:
             message: "System memory usage of {{ $value | humanize }} on {{ $labels.node }} exceeds 90% of the reservation. Reserved memory ensures system processes can function even when the node is fully allocated and protects against workload out of memory events impacting the proper functioning of the node. The default reservation is expected to be sufficient for most configurations and should be increased (https://docs.openshift.com/container-platform/latest/nodes/nodes/nodes-nodes-managing.html) when running nodes with high numbers of pods (either due to rate of change or at steady state)."
+    - name: containers-oom-killed-warning
+      rules:
+        - alert: ContainersOOMKilledWarning
+          expr: |
+            sum by (node) (container_runtime_crio_containers_oom_total) > 1
+          for: 15m
+          labels:
+            severity: warning
+          annotations:
+            message: "{{ $value | humanize }} containers have been killed on {{ $labels.node }} because they ran out of memory (OOM). Please make sure to limit the memory for the containers to not overcommit the node or try to reduce the total number of pods per node."
+    - name: containers-oom-killed-critical
+      rules:
+        - alert: ContainersOOMKilledCritical
+          expr: |
+            sum by (node) (container_runtime_crio_containers_oom_total) > 5
+          for: 15m
```

Comment:
Is there a reason to wait 15 min here to alert? We don't seem to be counting the number of OOM containers which would make sense to wait for but the number of killed containers which won't change even if recovered. So trying to figure out what the delay accomplishes? Or am I misunderstanding the alert?

Reply:
This alert is an a replacement for the existing alert `MultipleContainersOOMKilled` which also waits for 15min. The expression for `MultipleContainersOOMKilled` looks like, 

```
sum(max by(namespace, container, pod) (increase(kube_pod_container_status_restarts_total[12m])) and max by(namespace, container, pod) (kube_pod_container_status_last_terminated_reason{reason="OOMKilled"}) == 1) > 5
```

@saschagrunert @haircommander How long crio metric `container_runtime_crio_containers_oom_total` will keep reporting OOMKilled that count after the container is OOMKilled? @kikisdeliveryservice has a point, that eventually `container_runtime_crio_containers_oom_total` should reduce the OOMKilled count after a certain timeout otherwise the recovered containers will keep getting counted in and the alert will keep on firing even after system comes out of memory pressure.

Reply:
I believe it is a global count for the time cri-o has been running. Is there persisted state for this alert? as in, can we check the amount that's been increased, or subtract the amount that we previously alerted on?

Reply:
I think we want to use `rate` here to smooth out when the node comes out of memory pressure.

Reply:
+1 to what @rphillips said, you definitely want to apply `rate()` (or `increase()`) to the counter metric otherwise your alert might continue to fire forever.

Reply:
@rphillips @simonpasquier Thanks for your input, I have updated the PR with rate()